webapp/src/js/services/rules-engine.js

```diff
@@ -42,14 +65,28 @@ angular.module('inboxServices').factory('RulesEngine', function(
   );
   let initialized = initialize();
 
+  const cancelDebounce = debounce => {
+    if (debounce) {
+      debounce.cancel();
+    }
+  };
+
+  const addContextualPropertyToTargets = settingsDoc => {
+    const items = settingsDoc.tasks && settingsDoc.tasks.targets && settingsDoc.tasks.targets.items || [];
+    items.forEach(item => {
+      item.isContextual = userDoc => item.context ? !!$parse(item.context)({ user: userDoc }) : true;
+    });
+    return settingsDoc;
```

Comment:
I'm really concerned about modifying the extending Settings doc with functions like this. I get what you're trying to achieve but if I read this correctly it's modifying the cached settings object which is a bad idea. It would be a little better if you cloned the settings object but it's still mixing plain JSON with Functions in a dangerous way.

Can you not just inject the `$parse` service when you `RulesEngineCore.initialize`?

Reply:
I started with that and it just cluttered the interfaces quite a lot. I can revisit tomorrow.

Reply:
I've revisisted and `initialize()` now takes a structured settings object instead of the settingsDoc. This gives rules-engine angular service the ability to filter contextual targets. It simplifies the interface to rules-engine by combing "options" and "settings" into one structured parameter.